core/src/main/java/org/pentaho/platform/dataaccess/datasource/api/resources/JDBCDatasourceResource.java

```diff
@@ -287,68 +278,38 @@ public Response getConnection( @PathParam( "name" ) String name ) {
+    @ResponseCode( code = 204, condition = "JDBC datasource added successfully." ),
+    @ResponseCode( code = 401, condition = "User is not authorized to add JDBC datasources." ),
     @ResponseCode( code = 304, condition = "Datasource was not modified" ),
-    @ResponseCode( code = 500, condition = "An unexected error occurred while adding the JDBC datasource." )
+    @ResponseCode( code = 500, condition = "An unexpected error occurred while adding the JDBC datasource." )
   } )
   public Response addOrUpdate( @PathParam( "connectionId" ) String connectionName, DatabaseConnection connection ) {
     try {
-      validateAccess();
+      DatasourceService.validateAccess();
       // Prefer the path name over the one in the DTO object
       connection.setId( connectionName );
-      IDatabaseConnection savedConn = null;
-      try {
-        savedConn = service.getConnectionByName( connectionName );
-      } catch ( ConnectionServiceException e ){
-        // unfortunatley getConnectionById throws an exception not returning null when the conneciton is not present.
 
-      } catch ( NullPointerException e){
-        // unfortunatley getConnectionById throws an exception not returning null when the conneciton is not present.
+      try {
+        IDatabaseConnection old = service.getConnectionByName( connectionName );
 
-      }
-      boolean success = false;
-      if( savedConn != null ) {
         if ( StringUtils.isBlank( connection.getPassword() ) ) {
-          connection.setPassword( savedConn.getPassword() );
+          connection.setPassword( old.getPassword() );
         }
-        connection.setId( savedConn.getId() );
-        success = service.updateConnection( connection );
-      } else {
-        success = service.addConnection( connection );
-      }
-
+        connection.setId( old.getId() );
 
-      if ( success ) {
-        return buildOkResponse();
-      } else {
-        return buildNotModifiedResponse();
+        return service.updateConnection( connection ) ? Response.noContent().build() : Response.notModified().build();
+      } catch ( ConnectionServiceException e ) {
+        // unfortunately getConnectionById throws an exception not returning null when the connection is not present.
+      } catch ( NullPointerException e ) {
+        // unfortunately getConnectionById throws an exception not returning null when the connection is not present.
       }
+
+      return service.addConnection( connection ) ? Response.noContent().build() : Response.notModified().build();
     } catch ( PentahoAccessControlException t ) {
-      throw new WebApplicationException( Response.Status.UNAUTHORIZED );
-    } catch ( Throwable t ){
+      return Response.status( Response.Status.UNAUTHORIZED ).build();
+    } catch ( Throwable t ) {
       logger.error( "Error " + t.getMessage() );
-      return buildServerErrorResponse();
+      return Response.serverError().build();
     }
   }
-
```

Comment:
All the following methods were created for unit test purposes only.
This is not necessary. See new unit tests.